auth/src/main/java/com/firebase/ui/auth/viewmodel/idp/LinkingProviderResponseHandler.java

```diff
@@ -33,11 +33,33 @@ protected void signIn(@NonNull AuthCredential credential,
         FirebaseUser currentUser = getCurrentUser();
         if (currentUser == null) {
             getAuth().signInWithCredential(credential)
-                    .addOnSuccessListener(new StartLink(response))
-                    .addOnFailureListener(new OnFailureListener() {
+                    .continueWithTask(new Continuation<AuthResult, Task<Void>>() {
                         @Override
-                        public void onFailure(@NonNull Exception e) {
-                            setResult(Resource.<IdpResponse>forFailure(e));
+                        public Task<Void> then(@NonNull Task<AuthResult> task) {
+                            if (mRequestedSignInCredential == null) {
+                                return Tasks.forResult(null);
+                            } else {
+                                return task.getResult().getUser()
+                                        .linkWithCredential(mRequestedSignInCredential)
+                                        .continueWith(new Continuation<AuthResult, Void>() {
+                                            @Override
+                                            public Void then(@NonNull Task<AuthResult> task) {
+                                                // Since we've already signed in, it's too late to
+                                                // backtrack so we just ignore any errors.
+                                                return null;
+                                            }
+                                        });
+                            }
+                        }
+                    })
+                    .addOnCompleteListener(new OnCompleteListener<Void>() {
+                        @Override
+                        public void onComplete(@NonNull Task<Void> task) {
+                            if (task.isSuccessful()) {
```

Comment:
Seems like the way you have this set up the task is guaranteed to be successful (in the `if` branch you go straight to success and in the `else` branch you drop errors).  Is that your intention?

I think this class would be improved with a healthy dose of comments since I am a little lost in what you're trying to achieve and why some error cases are ignored.

Reply:
See above, my B.